src/stream/stream/peekable.rs

```diff
@@ -0,0 +1,66 @@
+use std::pin::Pin;
+
+use crate::future::Future;
+use crate::stream::Stream;
+use crate::task::{Context, Poll};
+use crate::stream::stream::StreamExt;
+
+#[doc(hidden)]
+#[allow(missing_debug_implementations)]
+pub struct Peekable<S: Stream>
+where
+    S: Sized,
+{
+    stream: S,
+    peeked: Option<Poll<Option<S::Item>>>,
+}
+
+impl<S> Peekable<S>
+where
+    S: Stream,
+{
+    pin_utils::unsafe_pinned!(stream: S);
+    pin_utils::unsafe_unpinned!(peeked: Option<Poll<Option<S::Item>>>);
+
+    pub(crate) fn new(stream: S) -> Self {
+        Peekable {
+            stream: stream,
+            peeked: None,
+        }
+    }
+
+    pub fn peek(mut self: Pin<&mut Self>) -> &Poll<Option<S::Item>> {
+        match &self.peeked {
+            Some(peeked) => &peeked,
+            None => {
+                // how to get the next `next` value? What about `Context`
```

Comment:
this is where I currently have troubles @yoshuawuyts , I have no idea how to get the next `next` value from here. Given that I store a `Option<Poll<Option<S::Item>>>` in the struct.